tools/cargos/src/main.rs

```diff
@@ -0,0 +1,254 @@
+    // Cargo doesn't know about the `rust-version` field, and so if we were to
+    // use Cargo with that toolchain, `pkg-meta` would return `null` when asked
+    // to retrieve the `rust-version` field. This also requires `RUSTFLAGS=''`
+    // to override any unstable `RUSTFLAGS` set by the caller.
+    let output = rustup(
+        ["run", "stable", "cargo", "metadata", "--format-version", "1"],
+        Some(("CARGO_TARGET_DIR", "target/cargos")),
+    );
+
+    let json = serde_json::from_slice::<Value>(&output.stdout).unwrap();
+    let packages = json.as_object().unwrap().get("packages").unwrap();
+    packages
+        .as_array()
+        .unwrap()
+        .iter()
+        .filter_map(|p| {
+            let package = p.as_object().unwrap();
+            if package.get("name").unwrap().as_str() == Some("zerocopy") {
+                let metadata = package.get("metadata").unwrap().as_object().unwrap();
+                let ci = metadata.get("ci").unwrap().as_object().unwrap();
+                Some(Versions {
+                    msrv: package.get("rust_version").unwrap().as_str().unwrap().to_string(),
+                    stable: ci.get("pinned-stable").unwrap().as_str().unwrap().to_string(),
+                    nightly: ci.get("pinned-nightly").unwrap().as_str().unwrap().to_string(),
+                    build_rs: metadata.get("build-rs").unwrap().as_object().unwrap().clone(),
+                })
+            } else {
+                None
+            }
+        })
+        .next()
+        .unwrap()
+}
+
+fn is_toolchain_installed(versions: &Versions, name: &str) -> Result<bool, Error> {
+    let version = versions.get(name)?;
+    let output = rustup(["run", version, "cargo", "version"], None);
+    if output.status.success() {
+        let output = rustup([&format!("+{version}"), "component", "list"], None);
+        let stdout = String::from_utf8(output.stdout).unwrap();
+        Ok(stdout.contains("rust-src (installed)"))
+    } else {
+        Ok(false)
+    }
+}
+
+fn install_toolchain_or_exit(versions: &Versions, name: &str) -> Result<(), Error> {
+    eprintln!("[cargos] missing either toolchain '{name}' or component 'rust-src'");
+    if env::vars().any(|v| v.0 == "GITHUB_RUN_ID") {
+        // If we're running in a GitHub action, then it's better to bail than to
+        // hang waiting for input we're never going to get.
+        process::exit(1);
+    }
+
+    loop {
+        eprint!("[cargos] would you like to install toolchain '{name}' and component 'rust-src' via 'rustup' (y/n)? ");
+        let mut input = [0];
+        io::stdin().read_exact(&mut input).unwrap();
+        match input[0] as char {
+            'y' | 'Y' => break,
+            'n' | 'N' => process::exit(1),
+            _ => (),
+        }
+    }
+
+    let version = versions.get(name)?;
+    Command::new("rustup")
+        .args(["toolchain", "install", &version, "-c", "rust-src"])
+        .status()
+        .unwrap();
```

Comment:
Could be worth factoring this out, but it can't be `fn rustup` (at least as it's written now) because that doesn't wire up stdin/stdout/stderr to the parent process's stdin/stdout/stderr.

Reply:
Done